### ISSUE TITLE
shell_exec security

### ISSUE THREAD

**Issue by bymayo** [spacecatninja/craft-imager-x#32]
We are trying to get webP images working on one of our clients sites, and noticed an error appear:

```
shell_exec() has been disabled for security reasons
```

(Sadly we've fixed the error so I can't provide the full trace!) But I did noticed it coming from the `vendor/spacecatninja/imager-x/src/transformers/CraftTransformer.php` file:

```
// Convert to webp with cwebp
$command = escapeshellcmd($config->getSetting('cwebpPath') . ' ' . $config->getSetting('cwebpOptions') . ' -q ' . $saveOptions['webp_quality'] . ' "' . $tempFile . '" -o "' . $path . '"');
$r = shell_exec($command);
```

I just wondered if this is a security issue running this command, as our hosting provider is very sceptical about enabling it.

### Additional info

- Imager version & edition:  v3.0.11 Pro
- Imager transformer: ?
- Craft version: 3.4.24
- PHP version: 7.2.31
- Image driver & version: GD
- Database driver & version: MySQL 5.7

**Comment by aelvan** [comment]
Although rare nowadays, there're some hosting providers that disable certain PHP commands to harden security, all commands that allow shell access being maybe the most common ones. 

Inherently, letting PHP execute shell commands opens you up to an attack vector where an intruder could be able to run commands on your server that he'd otherwise need shell access to run. This requires the intruder to have some way of manipulating the command being run, for instance by supplying a request parameter that is picked up by PHP and used to execute shell commands. Or a vulnerability where he can inject PHP to be run on the server (as was the case with the SEOmatic vulnerability that was discovered lately).

How much damage an intruder can do, depends largely on how much access the PHP process have. Most of the time, an intruder would "only" be able to wreck havoc to your website, but in a shared environment a badly configured server/webserver could also let the intruder attack horizontally to other sites and vertically to other parts of the server. 

In the case of Imager, no user-supplied parameters are added to the commands being run, and all commands are run through [`escapeshellcmd`](https://www.php.net/manual/en/function.escapeshellcmd.php) which escapes characters that are used to exit the command being run (which is what an intruder wants). If an intruder is able to manipulate the config file, or upload a file with a malicious filename that is later transformed (pretty sure that's only theoretically possible, as Craft/Yii/OS would catch that), there is a possibility that Imager would execute unwanted commands. But in that case, the server has been penetrated in other ways, and it's probably easier for the intruder to just execute shell commands directly. 

So, bottom line... Allowing/blocking the use of shell_exec _is a valid security concern_. I'm pretty confident that Imager wouldn't open you up to exploits using this as Imager doesn't rely on any user-provided input - but vulnerabilities in other scripts could. That said, if your server/webserver is properly configured, it wouldn't make a huge difference if the command was enabled or not. _It's the vulnerability in the code that's the problem_.

Let me know if that answers your question and if you want me to elaborate on something. :)

**Comment by bymayo** [comment]
@aelvan Wow, thanks for that detailed reply! I didn't expect there to be an issue, I just wondered if you knew of any reason why this was being used really. We've only had 2 hosting companies question it, and when I come to think about it, it's usually ones that have multiple sites on the server, not just one.

I can sleep now tonight :)
